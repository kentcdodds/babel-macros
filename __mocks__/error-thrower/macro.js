// const printAST = require('ast-pretty-print')
const {createMacro} = require('../../src')

module.exports = createMacro(evalMacro)

function evalMacro() {
  throw new Error('not helpful')
}
