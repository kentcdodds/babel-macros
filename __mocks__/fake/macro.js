// this is used to make sure that you can require macro from node_modules
module.exports = jest.fn()
