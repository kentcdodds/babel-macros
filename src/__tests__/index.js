import path from 'path'
// eslint-disable-next-line
import fakeMacro from 'fake/macro'
import babel from 'babel-core'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

afterEach(() => {
  fakeMacro.mockClear()
})

const projectRoot = path.join(__dirname, '../../')

// we don't actually need this,
// but I found that it removes double escaping when I include it...
expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename, parserOpts: {plugins: ['jsx']}},
  tests: [
    {
      title: 'does nothing to code that does not import macro',
      snapshot: false,
      code: `
        import foo from './some-file-without-macro';
        const bar = require('./some-other-file-without-macro');
      `,
    },
    {
      title: 'does nothing but remove macros if it is unused',
      code: `
        import foo from './some-macros-that-doesnt-even-need-to-exist.macro'
        export default 'something else'
      `,
    },
    {
      title: 'works with import',
      code: `
        import myEval from './fixtures/eval.macro'
        const x = myEval\`34 + 45\`
      `,
    },
    {
      title: 'works with require',
      code: `
        const evaler = require('./fixtures/eval.macro')
        const x = evaler\`34 + 45\`
      `,
    },
    {
      title: 'works with function calls',
      code: `
        import myEval from './fixtures/eval.macro'
        const x = myEval('34 + 45')
      `,
    },
    {
      title: 'Works as a JSXElement',
      code: `
        import MyEval from './fixtures/eval.macro'
        const x = <MyEval>34 + 45</MyEval>
      `,
    },
    {
      title: 'Supports named imports',
      code: `
        import {css as CSS, styled as STYLED} from './fixtures/emotion.macro'
        const red = CSS\`
          background-color: red;
        \`

        const Div = STYLED.div\`
          composes: \${red}
          color: blue;
        \`
      `,
    },
    {
      title: 'supports macros from node_modules',
      code: `
        import fakeMacro from 'fake/macro'
        fakeMacro('hi')
      `,
      teardown() {
        // kinda abusing the babel-plugin-tester API here
        // to make an extra assertion
        expect(fakeMacro).toHaveBeenCalledTimes(1)
        expect(fakeMacro).toHaveBeenCalledWith({
          references: expect.any(Object),
          state: expect.any(Object),
          babel: expect.any(Object),
          isBabelMacrosCall: true,
        })
        expect(fakeMacro.mock.calls[0].babel).toBe(babel)
      },
    },
    {
      title: 'forwards MacroErrors thrown by the macro',
      error: true,
      code: `
        import errorThrower from './fixtures/macro-error-thrower.macro'
        errorThrower('hey')
      `,
    },
    {
      title: 'prepends the relative path for errors thrown by the macro',
      error: true,
      code: `
        import errorThrower from './fixtures/error-thrower.macro'
        errorThrower('hey')
      `,
    },
    {
      title: 'appends the npm URL for errors thrown by node modules',
      error: true,
      code: `
        import errorThrower from 'error-thrower.macro'
        errorThrower('hi')
      `,
    },
  ],
})

test('throws error if it is not transpiled', () => {
  const untranspiledMacro = plugin.createMacro(() => {})
  expect(() =>
    untranspiledMacro({source: 'untranspiled.macro'}),
  ).toThrowErrorMatchingSnapshot()
})
