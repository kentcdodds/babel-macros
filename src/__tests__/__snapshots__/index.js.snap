// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros Supports named imports: Supports named imports 1`] = `

import {css as CSS, styled as STYLED} from './fixtures/emotion.macro'
const red = CSS\`
  background-color: red;
\`

const Div = STYLED.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = "background-color: red;";
const Div = STYLED.div\`undefined\`;

`;

exports[`macros Works as a JSXElement: Works as a JSXElement 1`] = `

import MyEval from './fixtures/eval.macro'
const x = <MyEval>34 + 45</MyEval>

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79;

`;

exports[`macros appends the npm URL for errors thrown by node modules with a slash: appends the npm URL for errors thrown by node modules with a slash 1`] = `

import errorThrower from 'babel-plugin-macros-test-error-thrower/macro'
errorThrower('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: babel-plugin-macros-test-error-thrower/macro: not helpful Learn more: https://www.npmjs.com/package/babel-plugin-macros-test-error-thrower

`;

exports[`macros appends the npm URL for errors thrown by node modules: appends the npm URL for errors thrown by node modules 1`] = `

import errorThrower from 'babel-plugin-macros-test-error-thrower.macro'
errorThrower('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: babel-plugin-macros-test-error-thrower.macro: not helpful Learn more: https://www.npmjs.com/package/babel-plugin-macros-test-error-thrower.macro

`;

exports[`macros does nothing but remove macros if it is unused: does nothing but remove macros if it is unused 1`] = `

import foo from './some-macros-that-doesnt-even-need-to-exist.macro'
export default 'something else'

      ↓ ↓ ↓ ↓ ↓ ↓

export default 'something else';

`;

exports[`macros forwards MacroErrors thrown by the macro: forwards MacroErrors thrown by the macro 1`] = `

import errorThrower from './fixtures/macro-error-thrower.macro'
errorThrower('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

MacroError: very helpful

`;

exports[`macros macros can set their configName and get their config: macros can set their configName and get their config 1`] = `

import configured from './configurable.macro'

configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

configured\`stuff\`;

`;

exports[`macros prepends the relative path for errors thrown by the macro: prepends the relative path for errors thrown by the macro 1`] = `

import errorThrower from './fixtures/error-thrower.macro'
errorThrower('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: ./fixtures/error-thrower.macro: very unhelpful

`;

exports[`macros supports compiled macros (\`__esModule\` + \`export default\`): supports compiled macros (\`__esModule\` + \`export default\`) 1`] = `

import {css, styled} from './fixtures/emotion-esm.macro'
const red = css\`
  background-color: red;
\`

const Div = styled.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = css\`
  background-color: red;
\`;
const Div = styled.div\`
  composes: \${red}
  color: blue;
\`;

`;

exports[`macros supports macros from node_modules: supports macros from node_modules 1`] = `

import fakeMacro from 'babel-plugin-macros-test-fake/macro'
fakeMacro('hi')

      ↓ ↓ ↓ ↓ ↓ ↓

fakeMacro('hi');

`;

exports[`macros throws an error if the macro is not properly wrapped: throws an error if the macro is not properly wrapped 1`] = `

import unwrapped from './fixtures/non-wrapped.macro'
unwrapped('hey')

      ↓ ↓ ↓ ↓ ↓ ↓

Error: The macro imported from "./fixtures/non-wrapped.macro" must be wrapped in "createMacro" which you can get from "babel-plugin-macros". Please refer to the documentation to see how to do this properly: https://github.com/kentcdodds/babel-plugin-macros/blob/master/other/docs/author.md#writing-a-macro

`;

exports[`macros when a plugin that replaces paths is used, macros still work properly: when a plugin that replaces paths is used, macros still work properly 1`] = `

import myEval from '../eval.macro'

const result = myEval\`+('4' + '2')\`

global.result = result

      ↓ ↓ ↓ ↓ ↓ ↓

const result = ("foobar", 42);
global.result = result;

`;

exports[`macros when there is an error reading the config, a helpful message is logged 1`] = `
Array [
  There was an error trying to load the config "configurableMacro" for the macro imported from "./configurable.macro. Please see the error thrown for more information.,
]
`;

exports[`macros when there is an error reading the config, a helpful message is logged: when there is an error reading the config, a helpful message is logged 1`] = `

import configured from './configurable.macro'

configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: this is a cosmiconfig error

`;

exports[`macros when there is no config to load, then no config is passed: when there is no config to load, then no config is passed 1`] = `

import configured from './configurable.macro'

configured\`stuff\`

      ↓ ↓ ↓ ↓ ↓ ↓

configured\`stuff\`;

`;

exports[`macros works with function calls: works with function calls 1`] = `

import myEval from './fixtures/eval.macro'
const x = myEval('34 + 45')

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79;

`;

exports[`macros works with import: works with import 1`] = `

import myEval from './fixtures/eval.macro'
const x = myEval\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79;

`;

exports[`macros works with require destructuring and aliasing: works with require destructuring and aliasing 1`] = `

const {css: CSS, styled: STYLED} = require('./fixtures/emotion.macro')
const red = CSS\`
  background-color: red;
\`

const Div = STYLED.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = "background-color: red;";
const Div = STYLED.div\`undefined\`;

`;

exports[`macros works with require destructuring: works with require destructuring 1`] = `

const {css, styled} = require('./fixtures/emotion.macro')
const red = css\`
  background-color: red;
\`

const Div = styled.div\`
  composes: \${red}
  color: blue;
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const red = "background-color: red;";
const Div = styled.div\`undefined\`;

`;

exports[`macros works with require: works with require 1`] = `

const evaler = require('./fixtures/eval.macro')
const x = evaler\`34 + 45\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 79;

`;
